pgjdbc/src/main/java/org/postgresql/jdbcurlresolver/JdbcUrlResolver.java

```diff
@@ -0,0 +1,326 @@
+/*
+ * Copyright (c) 2022, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.jdbcurlresolver;
+
+import static org.postgresql.jdbcurlresolver.JdbcUrlParser.saveProperty;
+import static org.postgresql.util.internal.Nullness.castNonNull;
+
+import org.postgresql.PGEnvironment;
+import org.postgresql.PGProperty;
+import org.postgresql.util.OSUtil;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Properties;
+import java.util.TreeMap;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+/**
+ * parser for jdbc URL.
+ * https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNSTRING
+ */
+public class JdbcUrlResolver {
+  private static final Logger LOGGER = Logger.getLogger(JdbcUrlResolver.class.getName());
+
+  // input
+  private final String url;
+  private final Properties defaults;
+
+  // to support unit tests
+  private @Nullable JdbcUrlResolverFatalException failException;
+
+  // to support unit tests
+  @Nullable JdbcUrlResolverFatalException getFailException() {
+    return failException;
+  }
+
+  /**
+   * @param url      jdbc url
+   * @param defaults defaults
+   */
+  public JdbcUrlResolver(String url, @Nullable Properties defaults) {
+    this.url = url;
+    this.defaults = defaults == null ? new Properties() : defaults;
+  }
+
+  /**
+   * Returns Properties
+   *
+   * @return result of parsing URL
+   */
+  public @Nullable Properties getResult() {
+    Properties result = new Properties();
+    try {
+      // override hierarchy
+      Properties p8 = new Properties(); // global defaults
```

Comment:
Seems to me that instead of using pN we should use the names that correspond to the comments

Reply:
For me, it gives the code clarity and readability.
p8 - priority 8 properties
p7 - priority 7 properties
...
Different names are not blocker for me. What do you prefer?
p8 -> globalDefaults
p7 -> orgPostgresDriverconfig
...

Reply:
I'd prefer names. Yes numbers provide priority, but names provide context